### PR TITLE
List Expressions - Paginated Api with Unit Test

### LINKED ISSUE #9: API: List Expressions - Paginated
GET /api/expressions?page=0

Returns a result set that includes the total records, the current page, and an array of expressions. Page size can be fixed at 25.

This endpoint does not require authentication.

### PR DESCRIPTION
**Pull request by ShadyNagy**
Fix #9

### CODE REVIEW

**Comment by ShadyNagy** [comment]
@antmdvs
What is your suggestion on this?
```
Returns a result set that includes the total records, the current page, and an array of expressions.
```

**Comment by antmdvs** [comment]
Is there a specific part of that requirement you're asking about or just generally how to do it?

**Comment by ShadyNagy** [comment]
I am thinking to create class Response and inside it, list and TotalRecords and CurrentPage, and inherited from it ExpressionResponse.
is that fine ? or there is something better?

**Comment by antmdvs** [comment]
That seems acceptable to me, though in the spirit of YAGNI, I'm not sure I would create a "Response" abstraction at this point -- having a single derived class would technically go against the [Reused Abstractions Principle](http://www.codemanship.co.uk/parlezuml/blog/?postid=934). Maybe just include the metadata in your `ExpressionResponse` for now? You could rename it to `PagedExpressionResponse` for clarity.

File: src/RegExLib.Web/Endpoints/Expressions/List.cs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+using RegExLib.SharedKernel.Interfaces;
+using Microsoft.AspNetCore.Mvc;
+using Swashbuckle.AspNetCore.Annotations;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace RegExLib.Web.Endpoints.Expressions
+{
+    public class List : BaseAsyncEndpoint<List<ExpressionResponse>>
```
Comment (antmdvs) [inline src/RegExLib.Web/Endpoints/Expressions/List.cs:12]: `BaseAsyncEndpoint<int, List<ExpressionResponse>>`

File: src/RegExLib.Web/Endpoints/Expressions/List.cs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        }
+
+        [HttpGet("/Expressions")]
+        [SwaggerOperation(
+            Summary = "Gets a list of all Expressions",
+            Description = "Gets a list of all Expressions",
+            OperationId = "Expression.List",
+            Tags = new[] { "ExpressionEndpoints" })
+        ]
+        public override async Task<ActionResult<List<ExpressionResponse>>> HandleAsync()
```
Comment (antmdvs) [inline src/RegExLib.Web/Endpoints/Expressions/List.cs:28]: `HandleAsync([FromQuery] int page)`

File: src/RegExLib.Web/Endpoints/Expressions/List.cs [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        {
+            var expressions = (await _repository.ListAsync<Expression>())
```
Comment (antmdvs) [inline src/RegExLib.Web/Endpoints/Expressions/List.cs:30]: `_repository.ListAsync<Expression>(page)`

File: tests/RegExLib.FunctionalTests/Api/Expressions/ApiExpressionsControllerList.cs [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+      var response = await _client.GetAsync("/api/expressions?page=0");
+      response.EnsureSuccessStatusCode();
+      var stringResponse = await response.Content.ReadAsStringAsync();
+      var result = JsonConvert.DeserializeObject<IEnumerable<Expression>>(stringResponse).ToList();
+
+      Assert.Single(result);
+      Assert.Contains(result, i => i.Title == ExpressionsSeed.Expression1.Title);
+    }
+  }
+}
```
Comment (antmdvs) [inline tests/RegExLib.FunctionalTests/Api/Expressions/ApiExpressionsControllerList.cs:34]: You might also consider testing these scenarios:

1. given _totalCount_ > _pageSize_ when _page_ = 0 then should return _pageSize_ elements
2. given totalCount > pageSize when _page_ = 1 then should return results for page 1
3. other boundary conditions
  a. _totalCount_ = 0
  a. _page_ < 0
  b. last page
  c. _page_ > # pages

File: RegExLib.v3.ncrunchsolution [hunk 5]
```diff
@@ -0,0 +1,6 @@
+﻿<SolutionConfiguration>
+  <Settings>
+    <AllowParallelTestExecution>False</AllowParallelTestExecution>
+    <SolutionConfigured>True</SolutionConfigured>
+  </Settings>
+</SolutionConfiguration>
```
Comment (antmdvs) [inline RegExLib.v3.ncrunchsolution:6]: Add *.ncrunchsolution to .gitignore?

File: src/RegExLib.Web/Endpoints/Expressions/List.cs [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+                .Select(expression => new ExpressionResponse(expression.Id, expression.Title, expression.Pattern, expression.Description, expression.AuthorId));
+
+            return Ok(expressions);
+        }
+    }
+}
```
Comment (antmdvs) [inline src/RegExLib.Web/Endpoints/Expressions/List.cs:36]: It looks like Steve added an .editorconfig to enforce formatting rules but this file doesn't conform. There might be others. I didn't check them all.

File: RegExLib.v3.ncrunchsolution (hunk 5 above)
Comment (ShadyNagy) [reply RegExLib.v3.ncrunchsolution:6]: this added by mistake. i will add it to .gitignore

File: src/RegExLib.Web/Endpoints/Expressions/List.cs (hunk 6 above)
Comment (ShadyNagy) [reply src/RegExLib.Web/Endpoints/Expressions/List.cs:36]: any file I work in I apply formatting rules.maybe this befoure i get latest.I will review all and apply.

File: tests/RegExLib.FunctionalTests/Api/Expressions/ApiExpressionsControllerList.cs (hunk 4 above)
Comment (ShadyNagy) [reply tests/RegExLib.FunctionalTests/Api/Expressions/ApiExpressionsControllerList.cs:34]: I already change seed class.
Do we need to add more than 25 records in expression seed to make these tests?

> 
> 
> You might also consider testing these scenarios:
> 
>     1. given _totalCount_ > _pageSize_ when _page_ = 0 then should return _pageSize_ elements
> 
>     2. given totalCount > pageSize when _page_ = 1 then should return results for page 1
> 
>     3. other boundary conditions
>        a. _totalCount_ = 0
>        a. _page_ < 0
>        b. last page
>        c. _page_ > # pages

**Comment by ShadyNagy** [comment]
@antmdvs 

What do you think about 1 class it is name Result and has all that we need.
like I did here (not all)
```csharp
public class Result<TData>
```
[Result.cs](https://github.com/ShadyNagy/Utilities/blob/master/src/ShadyNagy.Utilities.Api/Result.cs)
so later we can add messages and error code also it will use for any class. I was made this to deal with Angular.

> 
> 
> Nice work. I think we need to decide whether or not we're using API Endpoints. If so, then I don't think we need the controller nor the DTO, right?

File: src/RegExLib.Web/Endpoints/Expressions/List.cs (hunk 6 above)
Comment (ShadyNagy) [reply src/RegExLib.Web/Endpoints/Expressions/List.cs:36]: Done [a40c02f](https://github.com/ardalis/RegExLib.com/pull/25/commits/a40c02fe19ad2b3bed0677f78aa3a41b2b36fba5)

File: RegExLib.v3.ncrunchsolution (hunk 5 above)
Comment (ShadyNagy) [reply RegExLib.v3.ncrunchsolution:6]: Done [0332028](https://github.com/ardalis/RegExLib.com/pull/25/commits/0332028396fef2d2075aef0035c528ba917f6c03)

**Comment by ardalis** [comment]
See also: https://github.com/ardalis/Result

But yes something that returns back metadata along with the set of data itself is the way to go. I don't think we need a generic solution just yet. A PagedExpressionResult type that was specific to the needs of this endpoint is probably sufficient for now. Start specific and only add abstraction and generic-ness when there's a need (YAGNI).

File: tests/RegExLib.FunctionalTests/Api/Expressions/ApiExpressionsControllerList.cs (hunk 4 above)
Comment (antmdvs) [reply tests/RegExLib.FunctionalTests/Api/Expressions/ApiExpressionsControllerList.cs:34]: I suppose there's a couple options:

1. If page size were made configurable then I think testing the degenerate case where pageSize = 1 would probably be sufficient.

2. If it stays fixed at 25, you would indeed need at least 26. For this option, I might consider generating the seed objects using something like `Enumerable.Range(1, perPage + 1).Select(index => new { FullName = $"FullName{index}"})`

**Pull request merged**